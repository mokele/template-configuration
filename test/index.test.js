const { TCCommand } = require('../src/')
const path = require('path')

let cmd
beforeEach(() => {
  cmd = new TCCommand()
})

describe('no readConfiguration function', () => {
  beforeEach(() => {
    cmd = new TCCommand()
    cmd.pushArguments(['aws', 'sts'])
  })

  test('getArguments throws', () => {
    expect(() => cmd.getArguments()).toThrow()
  })
})

describe('init', () => {
  beforeEach(() => {
    cmd = new TCCommand()
    cmd.pushArguments(['init'])
  })

  test('isOwnCommand', () => {
    expect(cmd.isOwnCommand()).toBe(true)
  })

  test('without existsConfigurationFunction', () => {
    expect(() => cmd.run()).toThrow()
  })

  describe('with existsConfigurationFunction => true', () => {
    let writeConfig, existsConfig
    beforeEach(() => {
      writeConfig = jest.fn()
      existsConfig = jest.fn().mockReturnValue(true)
      cmd.existsConfiguration(existsConfig)
      cmd.writeConfiguration(writeConfig)
    })

    test('throws file exist', () => {
      expect(() => cmd.run()).toThrow()
      expect(existsConfig).toHaveBeenCalled()
    })

    test('does not call writeConfiguration', () => {
      expect(writeConfig).not.toHaveBeenCalled()
    })
  })

  describe('with existsConfigurationFunction => false', () => {
    let existsConfig
    beforeEach(() => {
      existsConfig = jest.fn().mockReturnValue(false)
      cmd.existsConfiguration(existsConfig)
    })

    test('without writeConfigurationFunction', () => {
      expect(() => cmd.run()).toThrow()
      expect(existsConfig).toHaveBeenCalled()
    })

    describe('with writeConfigurationFile', () => {
      let writeConfig
      beforeEach(() => {
        writeConfig = jest.fn()
        cmd.writeConfiguration(writeConfig)
        cmd.run()
      })

      test('calls writeConfig', () => {
        expect(writeConfig).toHaveBeenCalled()
      })
    })
  })
})

describe('with readConfigution function', () => {
  let readConfig
  beforeEach(() => {
    readConfig = jest.fn().mockReturnValue(JSON.stringify({
      Parameters: {
        key1: 'value1',
        key2: 'value2'
      },
      Tags: {
        tagKey1: 'tagValue1',
        tagKey2: 'tagValue2'
      }
    }))
    cmd.readConfiguration(readConfig)
  })

  describe('--dryrun', () => {
    let log
    beforeEach(() => {
      log = jest.fn()
      cmd.setLogger(log)
      cmd.pushArguments(['--dryrun', 'aws', 'sts'])
      cmd.getArguments()
    })

    test('isDryRun is true', () => {
      expect(cmd.isDryRun()).toBe(true)
    })

    test('logged command', () => {
      expect(log).toHaveBeenCalledWith('aws sts')
    })
  })

  describe('--debug', () => {
    let log
    beforeEach(() => {
      log = jest.fn()
      cmd.setLogger(log)
      cmd.pushArguments(['--debug', 'aws', 'sts'])
      cmd.getArguments()
    })

    test('isDryRun is false', () => {
      expect(cmd.isDryRun()).toBe(false)
    })

    test('logged command', () => {
      expect(log).toHaveBeenCalledWith('aws sts')
    })
  })

  describe('--config', () => {
    beforeEach(() => {
      cmd.pushArguments(['--config', path.join('template-configuration', 'test.json')])
      cmd.pushArguments(['aws', 'cloudformation', 'deploy'])
      cmd.getArguments()
    })

    test('readConfig called with file argument', () => {
      expect(readConfig).toHaveBeenCalledWith(path.resolve(path.join('template-configuration', 'test.json')))
    })
  })

  describe('aws', () => {
    beforeEach(() => {
      cmd.pushArgument('aws')
    })

    describe('non-parameter command', () => {
      beforeEach(() => {
        cmd.pushArgument('sts')
      })

      test('generates command with parameters', () => {
        expect(cmd.getArguments()).toMatchSnapshot()
      })
    })

    describe('cloudformation', () => {
      beforeEach(() => {
        cmd.pushArgument('cloudformation')
      })

      describe('deploy', () => {
        beforeEach(() => {
          cmd.pushArgument('deploy')
        })

        test('generates command with parameters', () => {
          expect(cmd.getArguments()).toMatchSnapshot()
        })

        describe('with help as last argument', () => {
          beforeEach(() => {
            cmd.pushArgument('help')
          })

          test('does not add parameters', () => {
            expect(cmd.getArguments()).toMatchSnapshot()
          })
        })
      })

      describe('create-change-set', () => {
        beforeEach(() => {
          cmd.pushArgument('create-change-set')
        })
        test('generates command with parameters', () => {
          expect(cmd.getArguments()).toMatchSnapshot()
        })
      })

      describe('create-change-set', () => {
        beforeEach(() => {
          cmd.pushArgument('create-change-set')
        })
        test('generates command with parameters', () => {
          expect(cmd.getArguments()).toMatchSnapshot()
        })
      })

      describe('create-stack', () => {
        beforeEach(() => {
          cmd.pushArgument('create-stack')
        })
        test('generates command with parameters', () => {
          expect(cmd.getArguments()).toMatchSnapshot()
        })
      })

      describe('create-stack-instances', () => {
        beforeEach(() => {
          cmd.pushArgument('create-stack-instances')
        })
        test('generates command with parameters', () => {
          expect(cmd.getArguments()).toMatchSnapshot()
        })
      })

      describe('create-stack-set', () => {
        beforeEach(() => {
          cmd.pushArgument('create-stack-set')
        })
        test('generates command with parameters', () => {
          expect(cmd.getArguments()).toMatchSnapshot()
        })
      })

      describe('update-stack', () => {
        beforeEach(() => {
          cmd.pushArgument('update-stack')
        })
        test('generates command with parameters', () => {
          expect(cmd.getArguments()).toMatchSnapshot()
        })
      })

      describe('update-stack-instances', () => {
        beforeEach(() => {
          cmd.pushArgument('update-stack-instances')
        })
        test('generates command with parameters', () => {
          expect(cmd.getArguments()).toMatchSnapshot()
        })
      })

      describe('update-stack-set', () => {
        beforeEach(() => {
          cmd.pushArgument('update-stack-set')
        })
        test('generates command with parameters', () => {
          expect(cmd.getArguments()).toMatchSnapshot()
        })
      })
    })

    describe('serverlessrepo', () => {
      beforeEach(() => {
        cmd.pushArgument('serverlessrepo')
      })

      describe('create-cloud-formation-change-set', () => {
        beforeEach(() => {
          cmd.pushArgument('create-cloud-formation-change-set')
        })

        test('generates command with parameters', () => {
          expect(cmd.getArguments()).toMatchSnapshot()
        })
      })
    })
  })

  describe('sam', () => {
    beforeEach(() => {
      cmd.pushArgument('sam')
    })
    describe('deploy', () => {
      beforeEach(() => {
        cmd.pushArgument('deploy')
      })

      test('generates command with parameters', () => {
        expect(cmd.getArguments()).toMatchSnapshot()
      })

      describe('with --help as last argument', () => {
        beforeEach(() => {
          cmd.pushArgument('--help')
        })

        test('does not add parameters', () => {
          expect(cmd.getArguments()).toMatchSnapshot()
        })
      })
    })
  })

  describe('rain', () => {
    beforeEach(() => {
      cmd.pushArgument('rain')
    })
    describe('deploy', () => {
      beforeEach(() => {
        cmd.pushArgument('deploy')
      })

      test('generates command with parameters', () => {
        expect(cmd.getArguments()).toMatchSnapshot()
      })

      describe('with --help as last argument', () => {
        beforeEach(() => {
          cmd.pushArgument('--help')
        })

        test('does not add parameters', () => {
          expect(cmd.getArguments()).toMatchSnapshot()
        })
      })

      describe('with -h as last argument', () => {
        beforeEach(() => {
          cmd.pushArgument('-h')
        })

        test('does not add parameters', () => {
          expect(cmd.getArguments()).toMatchSnapshot()
        })
      })
    })
  })
})
