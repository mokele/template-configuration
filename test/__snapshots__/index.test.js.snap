// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with readConfigution function aws cloudformation create-change-set generates command with parameters 1`] = `
Array [
  "aws",
  "cloudformation",
  "create-change-set",
  "--parameters",
  "ParameterKey=key1,ParameterValue=value1",
  "ParameterKey=key2,ParameterValue=value2",
  "--tags",
  "Key=tagKey1,Value=tagValue1",
  "Key=tagKey2,Value=tagValue2",
]
`;

exports[`with readConfigution function aws cloudformation create-change-set generates command with parameters 2`] = `
Array [
  "aws",
  "cloudformation",
  "create-change-set",
  "--parameters",
  "ParameterKey=key1,ParameterValue=value1",
  "ParameterKey=key2,ParameterValue=value2",
  "--tags",
  "Key=tagKey1,Value=tagValue1",
  "Key=tagKey2,Value=tagValue2",
]
`;

exports[`with readConfigution function aws cloudformation create-stack generates command with parameters 1`] = `
Array [
  "aws",
  "cloudformation",
  "create-stack",
  "--parameters",
  "ParameterKey=key1,ParameterValue=value1",
  "ParameterKey=key2,ParameterValue=value2",
  "--tags",
  "Key=tagKey1,Value=tagValue1",
  "Key=tagKey2,Value=tagValue2",
]
`;

exports[`with readConfigution function aws cloudformation create-stack-instances generates command with parameters 1`] = `
Array [
  "aws",
  "cloudformation",
  "create-stack-instances",
  "--parameter-overrides",
  "ParameterKey=key1,ParameterValue=value1",
  "ParameterKey=key2,ParameterValue=value2",
]
`;

exports[`with readConfigution function aws cloudformation create-stack-set generates command with parameters 1`] = `
Array [
  "aws",
  "cloudformation",
  "create-stack-set",
  "--parameters",
  "ParameterKey=key1,ParameterValue=value1",
  "ParameterKey=key2,ParameterValue=value2",
  "--tags",
  "Key=tagKey1,Value=tagValue1",
  "Key=tagKey2,Value=tagValue2",
]
`;

exports[`with readConfigution function aws cloudformation deploy generates command with parameters 1`] = `
Array [
  "aws",
  "cloudformation",
  "deploy",
  "--parameter-overrides",
  "key1=value1",
  "key2=value2",
  "--tags",
  "tagKey1=tagValue1",
  "tagKey2=tagValue2",
]
`;

exports[`with readConfigution function aws cloudformation deploy with help as last argument does not add parameters 1`] = `
Array [
  "aws",
  "cloudformation",
  "deploy",
  "help",
]
`;

exports[`with readConfigution function aws cloudformation update-stack generates command with parameters 1`] = `
Array [
  "aws",
  "cloudformation",
  "update-stack",
  "--parameters",
  "ParameterKey=key1,ParameterValue=value1",
  "ParameterKey=key2,ParameterValue=value2",
  "--tags",
  "Key=tagKey1,Value=tagValue1",
  "Key=tagKey2,Value=tagValue2",
]
`;

exports[`with readConfigution function aws cloudformation update-stack-instances generates command with parameters 1`] = `
Array [
  "aws",
  "cloudformation",
  "update-stack-instances",
  "--parameter-overrides",
  "ParameterKey=key1,ParameterValue=value1",
  "ParameterKey=key2,ParameterValue=value2",
]
`;

exports[`with readConfigution function aws cloudformation update-stack-set generates command with parameters 1`] = `
Array [
  "aws",
  "cloudformation",
  "update-stack-set",
  "--parameters",
  "ParameterKey=key1,ParameterValue=value1",
  "ParameterKey=key2,ParameterValue=value2",
  "--tags",
  "Key=tagKey1,Value=tagValue1",
  "Key=tagKey2,Value=tagValue2",
]
`;

exports[`with readConfigution function aws non-parameter command generates command with parameters 1`] = `
Array [
  "aws",
  "sts",
]
`;

exports[`with readConfigution function rain deploy generates command with parameters 1`] = `
Array [
  "rain",
  "deploy",
  "--params",
  "key1=value1,key2=value2",
  "--tags",
  "tagKey1=tagValue1,tagKey2=tagValue2",
]
`;

exports[`with readConfigution function rain deploy with --help as last argument does not add parameters 1`] = `
Array [
  "rain",
  "deploy",
  "--help",
]
`;

exports[`with readConfigution function rain deploy with -h as last argument does not add parameters 1`] = `
Array [
  "rain",
  "deploy",
  "-h",
]
`;

exports[`with readConfigution function sam deploy generates command with parameters 1`] = `
Array [
  "sam",
  "deploy",
  "--parameter-overrides",
  "ParameterKey=key1,ParameterValue=value1",
  "ParameterKey=key2,ParameterValue=value2",
  "--tags",
  "Key=tagKey1,Value=tagValue1",
  "Key=tagKey2,Value=tagValue2",
]
`;

exports[`with readConfigution function sam deploy with --help as last argument does not add parameters 1`] = `
Array [
  "sam",
  "deploy",
  "--help",
]
`;
